### PR TITLE
feat: extended functional api of nanmin

### LINKED ISSUE #26130: nanmin

### PR DESCRIPTION
**Pull request by Aaryan562**
Close #26130

### CODE REVIEW

File: ivy/data_classes/array/experimental/statistical.py [hunk 1]
```diff
@@ -180,6 +180,25 @@ def nanmean(
...
+        self: ivy.Array,
+        /,
+        *,
+        axis: Optional[Union[Tuple[int], int]] = None,
+        out: Optional[ivy.Array] = None,
+        keepdims: Optional[bool] = False,
+        initial: Optional[Union[int, float, complex]] = None,
+        where: Optional[ivy.Array] = None,
+    ) -> ivy.Array:
+        return ivy.nanmin(
```
Comment (HaiderSultanArc) [inline ivy/data_classes/array/experimental/statistical.py:193]: You have to add Doc Strings as well.

File: ivy/data_classes/container/experimental/statistical.py [hunk 2]
```diff
@@ -444,6 +444,35 @@ def nanmean(
...
+        keepdims: Optional[Union[bool, ivy.Container]] = False,
+        key_chains: Optional[Union[List[str], Dict[str, str]]] = None,
+        to_apply: bool = True,
+        prune_unapplied: bool = False,
+        map_sequences: bool = False,
+        out: Optional[ivy.Array] = None,
+        initial: Optional[Union[int, float, complex, ivy.Container]] = 1,
+        where: Optional[Union[ivy.Array, ivy.Container]] = None,
+    ) -> ivy.Container:
+        return ContainerBase.multi_map_in_static_method(
```
Comment (HaiderSultanArc) [inline ivy/data_classes/container/experimental/statistical.py:462]: Doc Strings here as well

File: ivy/functional/backends/jax/experimental/statistical.py [hunk 3]
```diff
@@ -162,6 +162,19 @@ def nanmean(
...
+    a: JaxArray,
+    /,
+    *,
+    axis: Optional[Union[int, Tuple[int]]] = None,
+    keepdims: Optional[bool] = False,
+    out: Optional[JaxArray] = None,
+    initial: Optional[Union[int, float, complex]] = None,
+    where: Optional[JaxArray] = None,
+) -> JaxArray:
+    return jnp.nanmin(a, axis=axis, keepdims=keepdims, initial=initial, out=out)
```
Comment (HaiderSultanArc) [inline ivy/functional/backends/jax/experimental/statistical.py:175]: you missed `where` parameter

File: ivy/functional/backends/jax/experimental/statistical.py [hunk 4]
```diff
@@ -162,6 +162,19 @@ def nanmean(
     return jnp.nanmean(a, axis=axis, keepdims=keepdims, dtype=dtype, out=out)
 
 
+def nanmin(
+    a: JaxArray,
+    /,
+    *,
+    axis: Optional[Union[int, Tuple[int]]] = None,
+    keepdims: Optional[bool] = False,
+    out: Optional[JaxArray] = None,
```
Comment (HaiderSultanArc) [inline ivy/functional/backends/jax/experimental/statistical.py:171]: I'd like `out` as the last argument just for consistency.

File: ivy/functional/backends/numpy/experimental/statistical.py [hunk 5]
```diff
@@ -167,6 +167,22 @@ def nanmean(
...
+    a: np.ndarray,
+    /,
+    *,
+    axis: Optional[Union[int, Tuple[int]]] = None,
+    keepdims: Optional[bool] = False,
+    initial: Optional[Union[int, float, complex]] = None,
+    where: Optional[np.ndarray] = None,
+    out: Optional[np.ndarray] = None,
+) -> np.ndarray:
+    return np.nanmin(a=a, axis=axis, keepdims=keepdims, initial=initial, out=out)
```
Comment (HaiderSultanArc) [inline ivy/functional/backends/numpy/experimental/statistical.py:180]: Missed `where` parameter here as well

File: ivy/functional/ivy/experimental/statistical.py [hunk 6]
```diff
@@ -253,6 +253,32 @@ def nanmean(
...
+    a: ivy.Array,
+    /,
+    *,
+    axis: Optional[Union[Tuple[int], int]] = None,
+    keepdims: Optional[bool] = False,
+    out: Optional[ivy.Array] = None,
+    initial: Optional[Union[int, float, complex]] = None,
+    where: Optional[ivy.Array] = None,
+) -> ivy.Array:
+    return ivy.current_backend(a).nanmin(
```
Comment (HaiderSultanArc) [inline ivy/functional/ivy/experimental/statistical.py:272]: Doc Strings for documentation here as well 😄

**Review by HaiderSultanArc — COMMENTED** [review]
You missed Paddle and MXNet Backends

**Comment by Aaryan562** [comment]
> You missed Paddle and MXNet Backends

Mxnet is not needed because none of the functions are been implemented in it. Paddle has been added.

**Review by HaiderSultanArc — COMMENTED** [review]
@Aaryan562 MXNet has `nanmin` in it's functional API: https://mxnet.apache.org/versions/master/api/python/docs/api/np/generated/mxnet.np.nanmin.html

**Comment by Aaryan562** [comment]
> @Aaryan562 MXNet has `nanmin` in it's functional API: https://mxnet.apache.org/versions/master/api/python/docs/api/np/generated/mxnet.np.nanmin.html
> 
> Even if it doesn't have the function it doesn't mean we leave the backend out. We don't provide a backend-specific implementation but our Functional API needs to handle that situation.
> 
> Read about types of Functions [here](https://unify.ai/docs/ivy/overview/deep_dive/function_types.html)

I mean to say when I opened the folder of mxnet backend none of the functions were been implemented so i thought that mxnet might not be the priority now. But since there is a functional api already so it would be relatively esy to create one.

**Comment by HaiderSultanArc** [comment]
🤔 @Aaryan562  yeah I guess let's skip MXNet for now. It's not testing the pipeline either.

**Comment by HaiderSultanArc** [comment]
@Aaryan562 you marked my other reviews but didn't update the code 😅

**Comment by Aaryan562** [comment]
> @Aaryan562 you marked my other reviews but didn't update the code 😅

I added all the docstrings and also the paddle backend and out and where arguments as you said. Did I miss any?

File: ivy/data_classes/container/experimental/statistical.py (hunk 2 above)
Comment (Aaryan562) [reply ivy/data_classes/container/experimental/statistical.py]: this is done

File: ivy/data_classes/array/experimental/statistical.py (hunk 1 above)
Comment (Aaryan562) [reply ivy/data_classes/array/experimental/statistical.py:229]: this is done

File: ivy/data_classes/container/experimental/statistical.py (hunk 2 above)
Comment (Aaryan562) [reply ivy/data_classes/container/experimental/statistical.py]: this is done

File: ivy/functional/backends/jax/experimental/statistical.py (hunk 3 above)
Comment (Aaryan562) [reply ivy/functional/backends/jax/experimental/statistical.py]: this is done

File: ivy/functional/backends/jax/experimental/statistical.py (hunk 4 above)
Comment (Aaryan562) [reply ivy/functional/backends/jax/experimental/statistical.py]: this is done

File: ivy/functional/backends/numpy/experimental/statistical.py (hunk 5 above)
Comment (Aaryan562) [reply ivy/functional/backends/numpy/experimental/statistical.py]: this is done

File: ivy/functional/ivy/experimental/statistical.py (hunk 6 above)
Comment (Aaryan562) [reply ivy/functional/ivy/experimental/statistical.py:308]: this is done

File: ivy/data_classes/container/experimental/statistical.py [hunk 7]
```diff
@@ -444,6 +444,130 @@ def nanmean(
             self, axis=axis, keepdims=keepdims, dtype=dtype, out=out
         )
 
+    @staticmethod
+    def static_nanmin(
```
Comment (HaiderSultanArc) [inline ivy/data_classes/container/experimental/statistical.py:448]: I think it's better if we make separate PRs for separate functions! 😄 Let's keep this one only for `nanmin`

**Review by HaiderSultanArc — CHANGES_REQUESTED** [review]
A lot of tests are failing as well!

File: ivy/data_classes/container/experimental/statistical.py (hunk 7 above)
Comment (Aaryan562) [reply ivy/data_classes/container/experimental/statistical.py:448]: It is for same function nanmin. If you check the file you will see for every function there is a static method created as mine.

Comment (HaiderSultanArc) [reply ivy/data_classes/container/experimental/statistical.py:448]: Sounds good!

**Comment by Aaryan562** [comment]
@HaiderSultanArc How should I remove the submodule array_api_test which have been pushed

**Comment by Aaryan562** [comment]
> A lot of tests are failing as well!

@HaiderSultanArc Except paddle all other backend the tests are passing 
image

File: ivy/functional/backends/paddle/experimental/statistical.py [hunk 8]
```diff
@@ -99,6 +99,34 @@ def _validate_quantile(q):
     return True
 
 
+@with_supported_dtypes(
+    {"2.5.1 and below": ("float32", "float64", "int32", "int64", "float16")},
```
Comment (HaiderSultanArc) [inline ivy/functional/backends/paddle/experimental/statistical.py:103]: float16 is not supported in Paddle. You should remove it! 😄

Comment (Aaryan562) [reply ivy/functional/backends/paddle/experimental/statistical.py]: > float16 is not supported in Paddle. You should remove it! 😄

This was not causing the error which I was getting but yes its true paddle dosent support float16. Passed all the test cases .
image

**Review by HaiderSultanArc — COMMENTED** [review]
Hey @Aaryan562 I can't see CI picking up your tests. Is your branch synced with upstream/main?

**Comment by Aaryan562** [comment]
> Hey @Aaryan562 I can't see CI picking up your tests. Is your branch synced with upstream/main?

yes it is. I update my branch every now and then to be in sync in with unify:main branch

File: ivy/data_classes/array/experimental/statistical.py [hunk 9]
```diff
@@ -180,6 +180,61 @@ def nanmean(
             self._data, axis=axis, keepdims=keepdims, dtype=dtype, out=out
         )
 
+    def nanmin(
+        self: ivy.Array,
+        /,
+        *,
+        axis: Optional[Union[Tuple[int], int]] = None,
+        out: Optional[ivy.Array] = None,
```
Comment (vedpatwardhan) [inline ivy/data_classes/array/experimental/statistical.py:188]: would be better if `out` was the last argument of the function, given that's what we've done with all other functions, same applies to the other instance and static methods and backend implementations

File: ivy/data_classes/container/experimental/statistical.py (hunk 7 above)
Comment (vedpatwardhan) [inline ivy/data_classes/container/experimental/statistical.py:448]: The container method should be private (`_static_nanmin`) like `_static_add` in `container/elementwise.py`

File: ivy/functional/backends/tensorflow/experimental/statistical.py [hunk 10]
```diff
@@ -66,6 +66,26 @@ def nanmean(
     return tf.experimental.numpy.nanmean(a, axis=axis, keepdims=keepdims, dtype=dtype)
 
 
+def nanmin(
+    a: Union[tf.Tensor, tf.Variable],
+    /,
+    *,
+    axis: Optional[Union[int, Tuple[int]]] = None,
+    keepdims: Optional[bool] = False,
+    out: Optional[Union[tf.Tensor, tf.Variable]] = None,
```
Comment (vedpatwardhan) [inline ivy/functional/backends/tensorflow/experimental/statistical.py:75]: different backend functions have been given different order of arguments (e.g. `out` is at the end in the `paddle` backend implementation but in the middle in the `tensorflow` backend implementation). Although this doesn't affect usage because all are keyword arguments, we should stick to a consistent function signature across the backends

File: ivy/functional/ivy/experimental/statistical.py [hunk 11]
```diff
@@ -253,6 +253,68 @@ def nanmean(
...
 
 
+@handle_out_argument
+@handle_nestable
+@handle_backend_invalid
+@handle_exceptions
+@to_native_arrays_and_back
+@handle_device
+def nanmin(
+    a: ivy.Array,
```
Comment (vedpatwardhan) [inline ivy/functional/ivy/experimental/statistical.py:263]: let's name `a` to `x` instead as that's the naming we've used for other statistical methods like `min`, `mean`, `std`, etc.

File: ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_statistical.py [hunk 12]
```diff
@@ -646,6 +646,37 @@ def test_nanmedian(
...
+        input_dtypes=[input_dtype],
+        test_flags=test_flags,
+        rtol_=1e-1,
+        atol_=1e-1,
+        backend_to_test=backend_fw,
+        fn_name=fn_name,
+        on_device=on_device,
+        a=x,
+        axis=axis,
+        keepdims=keep_dims,
```
Comment (vedpatwardhan) [inline ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_statistical.py:676]: we should be generating values for the `initial` and `where` arguments as well in the tests

**Review by vedpatwardhan — CHANGES_REQUESTED** [review]
Just requested a few changes as discussed, thanks @Aaryan562 😄
Regarding the CI not picking up this function, @RashulChutani could you please take a look if the issue is due to the mappings changing too much since the last time we recallibrated? Thanks 😄

File: ivy/data_classes/container/experimental/statistical.py (hunk 7 above)
Comment (Aaryan562) [reply ivy/data_classes/container/experimental/statistical.py:448]: @vedpatwardhan In the container file the other functions are not private like nanmean,nanprod etc. I mean why only this nanmin function should be private? In the container/elementwise.py all the functions are private.

File: ivy/functional/ivy/experimental/statistical.py (hunk 11 above)
Comment (Aaryan562) [reply ivy/functional/ivy/experimental/statistical.py:263]: @vedpatwardhan there are other functions in the file which are defined as a. So should I change them as well?

File: ivy/data_classes/container/experimental/statistical.py (hunk 7 above)
Comment (vedpatwardhan) [reply ivy/data_classes/container/experimental/statistical.py:448]: ideally they should be private as well, it must've gotten added as public by mistake. In general all container static methods are supposed to be private 😄

File: ivy/functional/ivy/experimental/statistical.py (hunk 11 above)
Comment (vedpatwardhan) [reply ivy/functional/ivy/experimental/statistical.py:263]: Yeah we could change the rest too, probably in a different PR, but `x` certainly seems more consistent with our API than `a` 😄

File: ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_statistical.py (hunk 12 above)
Comment (Aaryan562) [reply ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_statistical.py:678]: I have generated initial values for initial but I am not sure of "where" as in the file
there are functions which take "where" arguments but the tests are not generated for it and nor there is any example in the backends of handling where.

File: ivy/functional/backends/tensorflow/experimental/statistical.py (hunk 10 above)
Comment (Aaryan562) [reply ivy/functional/backends/tensorflow/experimental/statistical.py]: Okay noted. Done in the latest commit

File: ivy/functional/ivy/experimental/statistical.py (hunk 11 above)
Comment (Aaryan562) [reply ivy/functional/ivy/experimental/statistical.py]: Okay sure noted.

File: ivy/data_classes/container/experimental/statistical.py (hunk 7 above)
Comment (Aaryan562) [reply ivy/data_classes/container/experimental/statistical.py:448]: > ideally they should be private as well, it must've gotten added as public by mistake. In general all container static methods are supposed to be private 😄

When I made it private I got some error regarding it. I dont remember the exact error but when I removed it all the test cases passed.

Comment (Aaryan562) [reply ivy/data_classes/container/experimental/statistical.py:448]: @vedpatwardhan AttributeError: 'ivy.data_classes.array.array' object has no attribute 'static_nanmin'
ivy/data_classes/container/base.py:4023: AttributeError
this is the error

Comment (Aaryan562) [reply ivy/data_classes/container/experimental/statistical.py]: > @vedpatwardhan AttributeError: 'ivy.data_classes.array.array' object has no attribute 'static_nanmin' ivy/data_classes/container/base.py:4023: AttributeError this is the error

kindly ignore fixed it. The container is made private in the latest commit

File: ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_statistical.py (hunk 12 above)
Comment (vedpatwardhan) [reply ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_statistical.py:678]: Basically we should be generating the `where` argument in those tests too. In general, every argument accepted by the function (except the `out` argument) should be generated in the data generation in the tests as well. For the `out` argument, it gets created inside the `test_function` anyway so we don't need to generate the `out` argument but we should generate every other argument.

Comment (Aaryan562) [reply ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_statistical.py:678]: > Basically we should be generating the `where` argument in those tests too. In general, every argument accepted by the function (except the `out` argument) should be generated in the data generation in the tests as well. For the `out` argument, it gets created inside the `test_function` anyway so we don't need to generate the `out` argument but we should generate every other argument.

@vedpatwardhan Can you give me an idea how should I generate the values of `where` argument and how to pass the test cases of it across the backends. Because I tried, but I am not sure how where argument actually works by reading the documentation.

Comment (HaiderSultanArc) [reply ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_statistical.py:678]: @Aaryan562 you can see the `where` argument in more detail [here](https://numpy.org/doc/stable/reference/generated/numpy.ufunc.reduce.html#numpy.ufunc.reduce)

File: ivy/data_classes/container/experimental/statistical.py [hunk 13]
```diff
@@ -444,6 +444,130 @@ def nanmean(
             self, axis=axis, keepdims=keepdims, dtype=dtype, out=out
         )
 
+    @staticmethod
+    def _static_nanmin(
+        input: ivy.Container,
```
Comment (vedpatwardhan) [inline ivy/data_classes/container/experimental/statistical.py:449]: `input` should also be named to `x`

File: ivy/functional/backends/jax/experimental/statistical.py [hunk 14]
```diff
@@ -162,6 +162,23 @@ def nanmean(
     return jnp.nanmean(a, axis=axis, keepdims=keepdims, dtype=dtype, out=out)
 
 
+def nanmin(
+    a: JaxArray,
```
Comment (vedpatwardhan) [inline ivy/functional/backends/jax/experimental/statistical.py:166]: `a` should be named to `x`

File: ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_statistical.py (hunk 12 above)
Comment (vedpatwardhan) [reply ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_statistical.py:678]: > @vedpatwardhan Can you give me an idea how should I generate the values of `where` argument and how to pass the test cases of it across the backends. Because I tried, but I am not sure how where argument actually works by reading the documentation. Also, there are no examples of `where` argument in the files through which I can take motivation

Basically the value of the `where` argument should be a boolean array broadcastable to `x`, there's helpers like `broadcastable_arrays` which could be used. @HaiderSultanArc could help out further 😄

**Review by vedpatwardhan — CHANGES_REQUESTED** [review]
Just requested a few more changes, thanks @Aaryan562 😄

**Comment by Aaryan562** [comment]
@vedpatwardhan @HaiderSultanArc I have added extra file test_nanmin.py just to show the code through which I am getting homogenous output across all backends when tested with falsifying example in hypothesis:
image

The output I got while testing manually is:
image

**Review by HaiderSultanArc — COMMENTED** [review]
It was failing for Paddle before. How did you resolve it? 🤔

**Comment by Aaryan562** [comment]
> It was failing for Paddle before. How did you resolve it? 🤔

Locally it is still failing for paddle when it take datatypes like float16 and bfloat16

**Comment by Aaryan562** [comment]
> It was failing for Paddle before. How did you resolve it? 🤔

What about the erroI have above, where I tested the following falsifying example manually it got correct results but on hypothesis it is failing

**Comment by vedpatwardhan** [comment]
Actually the example that's failing in the tests has `container=True` and `keepdims=True`. I assume we're passing `keepdims=True` and also passing a `ivy.Container` for `x` when testing manually? Thanks @Aaryan562 😄

**Comment by Aaryan562** [comment]
> Actually the example that's failing in the tests has `container=True` and `keepdims=True`. I assume we're passing `keepdims=True` and also passing a `ivy.Container` for `x` when testing manually? Thanks @Aaryan562 😄

@vedpatwardhan Oh yes, you are right. When passing container, it fails. Thanks for debugging. Fixed it!
Also, since i am using `float` in `helpers.get_dtypes` it is also generating `float16` and `bfloat16` datatypes which are making paddle backend to fail. Could you tell me how to go about it?

File: ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_statistical.py (hunk 12 above)
Comment (Aaryan562) [reply ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_statistical.py:681]: @vedpatwardhan I searched in the docs but I did not find any helpers like `broadcastable_arrays` there are `array_and_broadcastable_shape` and `array_bools`

Comment (vedpatwardhan) [reply ivy_tests/test_ivy/test_functional/test_experimental/test_core/test_statistical.py:681]: `broadcastable_arrays` is present in `test_core/test_dtype.py`, `mutually_broadcastable_shapes` is present in `helpers/hypothesis_helpers/array_helpers.py`. Not sure why they don't show up in the docs

**Comment by vedpatwardhan** [comment]
I think the above error is occurring because the `where` argument is a `float` whereas the `where` argument should be a `bool` array. I see that you've generated a `bool` array for `where` in the `handle_test` decorator but while calling the `test_function` it gets converted to `float`. This is because the `input_dtypes` passed to the `test_function` is a `['float32']` which is considered as the dtype for all the arrays passed to the function (which includes `x` and `where`). So to solve this issue the `input_dtypes` passed to the function should be a `['float32', 'bool']` 😄

**Comment by Aaryan562** [comment]
@vedpatwardhan @HaiderSultanArc All the test cases ran succesfully
image

**Review by HaiderSultanArc — COMMENTED** [review]
@vedpatwardhan CI is not picking these tests either.

**Comment by vedpatwardhan** [comment]
> @vedpatwardhan CI is not picking these tests either.

Not sure what the problem is, I'll try to recallibrate the mapping again this weekend to see if that fixes the issue. Thanks @HaiderSultanArc 😄

**Review by vedpatwardhan — APPROVED** [review]
lgtm! We can get the PR merged based on what @HaiderSultanArc thinks. Thanks @Aaryan562 😄

**Comment by Aaryan562** [comment]
@vedpatwardhan @HaiderSultanArc Thank you for patiently reviewing my code, learnt a lot. Onto the next one!
